phoenix-core-server/src/main/java/org/apache/phoenix/replication/ReplicationLogGroup.java

```diff
@@ -378,38 +719,408 @@ protected MetricsReplicationLogGroupSource createMetricsSource() {
+        }
+
+        /**
+         * Processes a single event from the Disruptor ring buffer. This method handles both data
+         * and sync events.
+         * <p>
+         * For data events, it:
+         * <ol>
+         *   <li>Sends the append event to the current mode.</li>
+         *   <li>Processes any pending syncs if this is the end of a batch.</li>
+         * </ol>
+         * <p>
+         * For sync events, it:
+         * <ol>
+         *   <li>Adds the sync future to the pending list.</li>
+         *   <li>Processes any pending syncs if this is the end of a batch.</li>
+         * </ol>
+         * <p>
+         * If an IOException occurs, it sends the failure event to the current replication mode
+         * and then switches to the new mode. After switching to the new mode, it replays the
+         * pending batch of un-synced appends and the failed event.
+         * </p>
+         *
+         * @param event The event to process
+         * @param sequence The sequence number of the event
+         * @param endOfBatch Whether this is the last event in the current batch
+         * @throws Exception if the operation fails after all retries
+         */
+        @Override
+        public void onEvent(LogEvent event, long sequence, boolean endOfBatch) throws Exception {
+            // Calculate time spent in ring buffer
+            long currentTimeNs = System.nanoTime();
+            long ringBufferTimeNs = currentTimeNs - event.timestampNs;
+            metrics.updateRingBufferTime(ringBufferTimeNs);
+            try {
+                switch (event.type) {
+                    case EVENT_TYPE_DATA:
+                        currentModeImpl.append(event.record);
+                        // Process any pending syncs at the end of batch.
+                        if (endOfBatch) {
+                            processPendingSyncs(sequence);
+                        }
+                        return;
+                    case EVENT_TYPE_SYNC:
+                        // Add this sync future to the pending list
+                        // OK, to add the same future multiple times when we rewind the batch
+                        // as completing an already completed future is a no-op
+                        pendingSyncFutures.add(event.syncFuture);
+                        // Process any pending syncs at the end of batch.
+                        if (endOfBatch) {
+                            processPendingSyncs(sequence);
+                        }
+                        // after a successful sync check the mode set on the replication group
+                        // Doing the mode check on sync points makes the implementation more robust
+                        // since we can guarantee that all unsynced appends have been flushed to the
+                        // replication log before we switch the replication mode
+                        ReplicationMode newMode = getMode();
```

Comment:
Mode change should be handled in at the beginning of processPendingSyncs() in order to make sure that it is not delayed indefinitely